CI: Port upload_to_owncloud.sh to python

Galvien opened the pull request:
Quoting from #1185:
 > As part of our efforts to clean up our scripts and workflow for certification tests, we need this to be refactored into a python script.

Therefore, port the `upload_to_owncloud.sh` script to Python3, resulting in a cleaner and more maintainable script.

Closes #1185

Commit:
ci/owncloud: add symlink to opts.py

in order to avoid multiple copies of the same boilerplate, add a
symbolic link to `test_flow.py`'s opts.py module, later to be used by
the owncloud script for printing.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
ci/owncloud: add a default webdav options file

although the script will utilize .netrc for getting the credentials
required for logging into owncloud, for whichever reason one might want,
it will also support getting the credentials from an external file.
this file serves as a skeleton. The format is identical to the python
dictionary utilized by the webdav module.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
ci/owncloud: create python script file

add a script file with opts and webdav import
and prepare for upcoming additions

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
ci/owncloud: add remote filesystem methods

in preparation for uploading files to the remote server, add methods for
recursively creating remote directories from a given path.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
ci/owncloud: add getter of server credentials

add a method for getting the webdav server's credentials, by default
they're retreived from ~/.netrc, but there's an option for retreiving
them from an external file.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
ci/owncloud: add upload method

Add a method uploading the contents of a local directory to a remote
webdav one, creating the remote directory if necessary.
The remote directory's tail is a unique uuid.
If the --direct arg is not given, the contents are firstly uploaded to a
temporary location in the webdav server.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
ci/owncloud: add main

add a pythonic main, parsing the arguments and preparing everything for
the upload() call

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
fixup! ci/owncloud: add getter of server credentials

tomereli commented on ci/owncloud/opts.py:
```diff
@@ -0,0 +1 @@
+../../tests/opts.py
```
I'm not sure I like this - why do you need symlink? Why can't you just import where you need opts.py with relative path?

Galvien replied:
See https://github.com/prplfoundation/prplMesh/pull/1302/files#r426802483

tomereli commented on ci/owncloud/owncloud_options.json:
```diff
@@ -0,0 +1,5 @@
+{
```
We definitly don't need more files for getting the credentials. Please allow only `~/.netrc`.

Galvien replied:
In your comment on #1185 you mentioned having an ability to support getting credentials from an external file.
By default the script uses `.netrc` credentials, but this leaves the ability to easily import credentials without digging through configurations.

arnout replied:
The comment you refer to says:
- it should be yaml, not json;
- it should be called owncloud_options.default.yml;
- if owncloud_options.yml exists, it should be used rather than the default (and it should be in .gitignore);
- credentials should *always* be taken from netrc so we don't have another file where we need to worry about visibility and permissions.

Nowhere does the comment say anything about getting credentials from a different file. It literally says "I disagree about using anything other than ~/.netrc".

tomereli commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,3 @@
+from webdav3.client import Client
+from opts import debug, message, opts
```
https://stackoverflow.com/questions/7505988/importing-from-a-relative-path-in-python
No need for a soft link.

Galvien replied:
From some reading I understand that implementing it the proper way would require adjustments to the filesystem (adding `__init__.py` files so that we have a package system, which is also kind of wrong, since what we're using is not package).

Changing the import to `from ...tests.opts import debug, message, opts` will result in an `ImportError` since there is no parent package.

There's a hacky way of doing it but it is highly discouraged by PEP.

A good solution IMO would be to implement a package of python utilities inside the prplMesh repository and accessing it with relative imports. But as it stands now, I would say using a symlink is the method that involves creating fewer files.

tomereli commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -1,3 +1,47 @@
+        try:
+            webdav_client.mkdir(p)
+        except BaseException as e:
+            debug(e)
```
Shouldn't we also raise the exception instead of hiding it?

Galvien replied:
Not exactly, since I'm pretty sure it might throw a 405 status code if the directory already exists, I got a few of those when running tests, so I did this just in case.
Looking at the source code though, It looks like `mkdir` doesn't raise an exception in the case of a 405, I didn't try running the code without the `try` block.

tomereli commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -1,8 +1,46 @@
+default_hostname = "https://ftp.essensium.com/owncloud/remote.php/dav/files/"
+
+
+def get_credentials(options_path=None, force_netrc=True):
```
As I said, this is redundant. There is absolutely no gain from having 2 different files for fetching credentials.
What you can do is add an optional param `netrc_path` which will default to `~/.netrc`.

Galvien replied:
Not exactly redundant as the `.netrc` file's hostanme field contains only a partial path, we could store only the `"owncloud/remote.php/dav/files/"` part, but in the shell version it's stored in the same way it was implemented here.

tomereli reviewed (approved):
have some small comments, but looking good so far assuming it works.

tomereli reviewed (changes requested):
I completely forgot - this is not complete until `gitlab-ci.yml` is updated to use the new script, so please add a commit which does that.

Commit:
ci: upload gitlab-ci.yml to use the py script

also, note that the python script's positional arguemnts are "local" and
then "remote", the other way around from the shell script, and more
in line with other apis.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

arnout commented on .gitlab-ci.yml:
```diff
@@ -152,11 +152,10 @@ upload-artifacts:
-      # "latest" afterwards:
-      ci/owncloud/upload_to_owncloud.sh -v "artifacts/latest" ./build
+      python3 ci/owncloud/upload_to_owncloud.py -v "artifacts/$CI_COMMIT_SHA-$CI_JOB_ID" ./build
+      # Copying takes a lot of time, and we want the "latest" folder to be updated
+      # "as fast as possible", so we make a copy and move it:
```
This comment should still be before the call to upload_to_owncloud, because it's about exactly that call.

arnout commented on ci/owncloud/owncloud_options.json:
```diff
@@ -0,0 +1,5 @@
+    "webdav_login": "Owncloud username",
+    "webdav_password": "Owncloud Password",
+    "webdav_hostname": "https://ftp.essensium.com/owncloud/remote.php/dav/files/"
+}
```
Please make sure the file ends with a single newline (no empty line either).

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+default_hostname = "https://ftp.essensium.com/owncloud/remote.php/dav/files/"
+
+
+def get_credentials(options_path=None, force_netrc=True):
```
Please don't implement features that are not used, unless there is a very clear plan to start using them in the near future. In this case, the two arguments are not used, so just remove them.

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+
+
+def get_credentials(options_path=None, force_netrc=True):
+    """
+    get_credentials(options_path=None, force_netrc=True)
+
+    Gets the user's default credentials from either an external options file or ~/.netrc
```
Don't repeat the function signature, and put the summary line on the same line as the `"""`.
```suggestion
    """Gets the user's default credentials from either an external options file or ~/.netrc
```

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    force_netrc : bool, optional
+                flag indicating whether to skip parsing the options file and instead directly
+                load the credentials from ~/.netrc
+    Returns
```
```suggestion

    Returns
```

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    dictionary with 3 key/value pairs,
+    keys are: {webdav_login", "webdav_password","webdav_hostname")
+    """
+    if not force_netrc and path.exists(options_path):
+        with open(options_path) as options_file:
+            options = json.load(options_file)
+            return options
```
Even when reading from netrc, the rest of the options should be taken from options_path.

Also, there should be a default for it.

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+                load the credentials from ~/.netrc
+    Returns
+    -------
+    dictionary with 3 key/value pairs,
+    keys are: {webdav_login", "webdav_password","webdav_hostname")
```
The dictionary will contain everything that is in the options file.

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+            return options
+
+    netrc_file = netrc.netrc()
+    credentials = netrc_file.authenticators("ftp.essensium.com")
```
It should take the hostname from the options that were parsed above.

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+import argparse
+from os import path
```
This is unusual - normally, you do just `import os` and use `os.path.foo()` in the code. Doing it differently is going to confuse people.

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+import argparse
+from os import path
+from getpass import getuser
```
Avoid `from x import y` if there's not a good reason for it. Here, getuser is used only once, so `getpass.getuser` is definitely not a problem.

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+from os import path
+from getpass import getuser
+import netrc
+from webdav3.client import Client
```
Here again: just `Client` is not very clear, it could be any client. `webdav3.client.Client` is much clearer.

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+
+
+def get_path_components(dirname):
+    """
+    get_path_components(dirname)
+
+    Gets a list of all subpaths needed to build a directory, including itself.
```
```suggestion
    """Gets a list of all subpaths needed to build a directory, including itself.
```

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    A python list of path strings
+    For an input 'a/b/c' returns ['a/b/c', 'a/b', 'a']
+    """
+    path_list = [dirname, ]
```
```suggestion
    path_list = [dirname]
```

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    Returns
+    -------
+    A python list of path strings
+    For an input 'a/b/c' returns ['a/b/c', 'a/b', 'a']
```
```suggestion
    For an input 'a/b/c' returns ['a', 'a/b', 'a/b/c']
```
At least, as far as I understand the code...

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    Parameters
+    ----------
+    dirname : str
+        A given directory path
+    Returns
```
```suggestion
        A given directory path

    Returns
```

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    """
+    path_list = [dirname, ]
+    tmp = path.split(dirname)
+    while tmp[0] != "" and not path.ismount(tmp[0]):
```
Since this is used for creating a path on the server, checking whether the local path is a mount point seems weird... Why do you do that?

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    For an input 'a/b/c' returns ['a/b/c', 'a/b', 'a']
+    """
+    path_list = [dirname, ]
+    tmp = path.split(dirname)
+    while tmp[0] != "" and not path.ismount(tmp[0]):
+        path_list.append(tmp[0])
+        tmp = path.split(tmp[0])
+    path_list.reverse()
+    return path_list
```
This seems overly complicated to me... How about:
```suggestion
    if not path.startswith('/'):
        raise ImplementationError("Only absolute paths are supported")
    components = dirname.split('/')
    pathlist = []
    previous = ''
    # First one is empty string - skip it
    for component in components[1:]:
        previous += '/' + component
        pathlist.append(previous)
    return pathlist
```

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+
+
+def upload(webdav_client, is_direct, remote_path, local_path, unique_id=None):
+    """
+    upload(webdav_client, is_direct, remote_path, local_path, unique_id)
+
+    Uploads the content of a directory to a directory in the remote webdav server
```
```suggestion
    """Upload the content of a directory to a directory in the remote webdav server.
```
Summary line should be imperative, not third person. And it should be a full sentence ending with .

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    message("""Uploading files from local directory {0} to remote directory {1},\n \
+         username is {2}, remote directory id is: {3}""".format(
+        local_path, remote_path, webdav_client.webdav.options["login"], webdav_client, unique_id))
+    remote_full_path = path.join(remote_path, unique_id)
```
I believe that in the original script, the unique_id was only added to the temporary path, not to the final destination? @rmelotte ?

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+        remote_temp_path = path.join("temp_uploads", unique_id)
+        mkdir_recursively(remote_temp_path)
+        webdav_client.upload_sync(remote_temp_path, local_path)
+        webdav_client.copy(remote_temp_path, remote_full_path)
```
Shouldn't this be move rather than copy?

Also, if the remote_full_path already exists, isn't it going to add a new subdirectory under remote_full_path, rather than copying/moving all the individual files and directories?

arnout commented on ci/owncloud/upload_to_owncloud.py:
```diff
@@ -0,0 +1,153 @@
+    args.remote_path = getattr(args, "remote-path")[0]
+    opts.verbose = args.verbose
+    webdav_options = get_credentials()
+    client = Client(webdav_options)
```
(nitpick) I would move the creation of the client inside the upload function, i.e. pass the dictionary as the first argument. That makes it easier to use this function by importing.
